syntax = "proto3";

service Fike {
    rpc read(ReadFileReq) returns (ReadFileRsp) {}
    rpc write(WriteFileReq) returns (WriteRsp) {}
}

message ReadFile {
    bytes buffer = 1; 
}

message ReadFileReq {
    string filename = 1;
    string chunkname = 2;
}

message WriteFileReq {
    string filename = 1;
    string chunkname = 2;
    bytes buffer = 3;
}

message WriteRsp {
}

service NameNodeService{
    rpc open(FileOpenReq) returns (stream DatanodeList) {}
    rpc create(FileCreateReq) returns (stream CreateRsp) {}
    rpc heart_beat(DatanodeInfo) return (HeartBeatRsp) {}
    rpc report(ChunkReport) return (Empty) {}
    rpc listin(Empty) returns (stream DirectoryContent) {}
    rpc get_followers(LeaderFollowersReq) returns (stream follower_info) {}
}

message FileOpenReq {
    string filename = 1;
}

message FileCreateReq {
    string filename = 1;
    int64 chunks_number = 2;
    string operation = 3;
}

message HeartBeatRsp{
    int64 cluster_id = 1;
    string id_datanode = 2;
    bool is_leader = 3;
}

message DatanodeList{
    string localization = 1;
    string chunkname = 2;
}

message WarningMessage{
    string message = 1;
}

message CreateRsp{
    oneof Response{
        DatanodeList datanode_list = 1;
        WarningMessage warning_message = 2;
    }
}

message Empty{}

message DatanodeInfo{
    string id = 1;
    string socket = 2;
    int64 cluster = 3;
}

message ChunkReport{
    string partname = 1;
    string filename = 2;
    string location = 3;
}

message DirectoryContent{
    string name = 1;
}

message LeaderFollowersReq{
    string leader_id = 1;
    int64 cluster_id = 2;
}

message follower_info{
    string follower_id = 1;
    string follower_location = 2;
}

